File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.cpp
class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int n=nums.size();
        int one=0;
        for(int i=0;i<n;i++){
            if(nums[i]==1){
                one++;
            }
        }
        if(one==0 || one==n){
            return 0;
        }
        int cur=0;
        for(int i=0;i<one;i++){
            if(nums[i]==1){
                cur++;
            }
        }
        int ans=one-cur;
        for(int i=1;i<n;i++){
            if(i+one<=n){
                if(nums[i+one-1]==1){
                    cur++;
                }
                if(nums[i-1]==1){
                    cur--;
                }
            }
            else{
                if(nums[i+one-n-1]==1){
                    cur++;
                }
                if(nums[i-1]==1){
                    cur--;
                }
            }
            ans=min(ans,one-cur);
            if(!ans){
                break;
            }
        }
        return ans;
    }
};

File: 2255-minimum-swaps-to-group-all-1s-together-ii/minimum-swaps-to-group-all-1s-together-ii.py
class Solution(object):
    def minSwaps(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        one=0
        for i in range(n):
            if nums[i]==1:
                one+=1

        if one==0 or one==n:
            return 0
    
        cur=0;
        for i in range(one):
            if nums[i]==1 :
                cur+=1
        
        ans=one-cur
        for i in range(1,n):
            if i+one<=n:
                if nums[i+one-1]==1:
                    cur+=1
        
                if nums[i-1]==1:
                    cur-=1

            else:
                if nums[i+one-n-1]==1:
                    cur+=1

                if nums[i-1]==1:
                    cur-=1

            ans=min(ans,one-cur)
            if not ans:
                break

        return ans;
        